### PR TITLE
[AArch64] Add patterns for constructive splice.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
SVE2 adds the constructive splice instruction, which takes a tuple.
Even though the register allocator must ensure that the tuple uses
consecutive registers for the tuple, it's likely to be more efficient
than using the destructive splice instruction when the first operand
is reused.

### CODE REVIEW

**Comment by aemerson** [comment]
According to this: https://developer.arm.com/documentation/ddi0602/2024-09/SVE-Instructions/SPLICE--Splice-two-vectors-under-predicate-control- the splice instructions require both SVE2 *and* SME, it looks like the splice definition currently is guarded under "SVE2OrSME".

**Comment by paulwalker-arm** [comment]
> According to this: https://developer.arm.com/documentation/ddi0602/2024-09/SVE-Instructions/SPLICE--Splice-two-vectors-under-predicate-control- the splice instructions require both SVE2 _and_ SME, it looks like the splice definition currently is guarded under "SVE2OrSME".

The specification says:
```
if !IsFeatureImplemented(FEAT_SVE2) && !IsFeatureImplemented(FEAT_SME) then
    EndOfDecode(Decode_UNDEF);
```
If both features are not implemented then it's a decode error.  This implies either of the features is sufficient to enable the instruction (along with a streaming mode check when targeting SME without SVE).

File: llvm/test/CodeGen/AArch64/sve-streaming-mode-fixed-length-concat.ll [hunk 1]
```diff
@@ -406,17 +406,33 @@ define void @concat_v8i64(ptr %a, ptr %b, ptr %c) {
...
+; SVE2-NEXT:    adrp x10, .LCPI15_1
+; SVE2-NEXT:    mov z2.h, w8
+; SVE2-NEXT:    ldr q3, [x9, :lo12:.LCPI15_0]
+; SVE2-NEXT:    ldr q4, [x10, :lo12:.LCPI15_1]
+; SVE2-NEXT:    ptrue p0.h, vl8
+; SVE2-NEXT:    // kill: def $d1 killed $d1 killed $z0_z1 def $z0_z1
+; SVE2-NEXT:    // kill: def $d0 killed $d0 killed $z0_z1 def $z0_z1
+; SVE2-NEXT:    mad z2.h, p0/m, z3.h, z4.h
+; SVE2-NEXT:    tbl z0.h, { z0.h, z1.h }, z2.h
+; SVE2-NEXT:    // kill: def $d0 killed $d0 killed $z0
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/sve-streaming-mode-fixed-length-concat.ll:422]: This change (regression?) looks strange. Not necessarily blocking this but do you know why we see a difference here between SVE2 and SME?

**Review by aemerson — COMMENTED** [review]
Yes, you're right.

File: llvm/test/CodeGen/AArch64/sve-streaming-mode-fixed-length-concat.ll (hunk 1 above)
Comment (aemerson) [reply llvm/test/CodeGen/AArch64/sve-streaming-mode-fixed-length-concat.ll:422]: Ok, can be a follow up fix then.

**Review by aemerson — APPROVED** [review]
(Not for this PR) but `ext` can also have a pattern added to use the constructive variant?

**Review by paulwalker-arm — APPROVED** [review]

**Pull request merged**